### PR TITLE
Ben/association task

### PR DESCRIPTION
**Pull request by benleetownsend**

### CODE REVIEW

**Comment by benleetownsend** [comment]
I've run the tests in this branch and sequence tests and everything passes. I will run the whole suite tonight just to verify, but should be okay.

File: finetune/association.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+from finetune.utils import indico_to_finetune_sequence, finetune_to_indico_sequence
+from finetune.input_pipeline import BasePipeline, ENCODER
+from finetune.errors import FinetuneError
+from finetune.base import LOGGER
+
+class AssociationPipeline(BasePipeline):
+    def __init__(self, config, multi_label):
+        super(AssociationPipeline, self).__init__(config)
+        self.multi_label = multi_label
+        self.association_encoder = SequenceLabelingEncoder()
```
Comment (madisonmay) [inline finetune/association.py]: Question -- should this be handled in a separate cls that inherits from SequenceLabelingEncoder and is returned by `_target_encoder`? Seems like it shouldn't live in the `__init__` of the pipeline.

File: finetune/association.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+        if self.config.possible_associations is None:
+            raise FinetuneError("Please set config.possible_associations before calling finetune.")
+        Xs, Y_new, association_type, association_idx, idxs = indico_to_finetune_sequence(
+            Xs, labels=Y, multi_label=False, none_value="<PAD>"
+        )
+
+        Y = list(zip(Y_new, association_type, association_idx, idxs)) if Y is not None else None
+        return super().finetune(Xs, Y=Y, batch_size=batch_size)
+
+    def prune_probs(self, prob_matrix, labels):
```
Comment (madisonmay) [inline finetune/association.py]: Should we renorm here too or are we only going to be extracting a max?

File: finetune/association.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+        Produces a list of most likely class labels as determined by the fine-tuned model.
+
+        :param X: A list / array of text, shape [batch]
+        :returns: list of class labels.
+        """
+        LOGGER.warning("config.viable_edges is not set, this is probably incorrect.")
+
+        #TODO(Ben) combine this into the sequence labeling model??
+
+        chunk_size = self.config.max_length - 2
```
Comment (madisonmay) [inline finetune/association.py]: Even if we don't factor everything out we probably should look into factoring out the chunking logic.

File: finetune/association.py [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+            return prob_matrix
+        for i, l1 in enumerate(labels):
+            if l1 not in viable_edges:
+                prob_matrix[i, :, :] = 0.0
+
+            elif None not in viable_edges[l1]:
+                prob_matrix[i, :, self.input_pipeline.association_pad_idx] = 0.0
+            for cls in possible_associations:
+                for j, l2 in enumerate(labels):
+                    if l1 not in viable_edges or l2 not in [c_t[0] for c_t in viable_edges[l1] if
```
Comment (madisonmay) [inline finetune/association.py]: Do we need to renorm at the end of this code for predict_proba purposes?

File: finetune/association.py [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+            for logits_i, trans_mat_i in zip(logits, trans_mats):
+                idx, prob = sequence_decode(logits_i, trans_mat_i)
+                label_idxs.append(idx)
+                label_probas.append(prob[:, :, 1:])
+            label_idxs = tf.stack(label_idxs, axis=-1)
+            label_probas = tf.stack(label_probas, axis=-1)
+        else:
+            label_idxs, label_probas = sequence_decode(logits, trans_mats)
+
+        association_prob = tf.nn.softmax(associations, axis=-1)
```
Comment (madisonmay) [inline finetune/association.py]: eventually replace with fancier decoding mechanism

File: finetune/base.py [hunk 6]
```diff
@@ -148,6 +149,7 @@ def finetune(self, Xs, Y=None, batch_size=None):
         batch_size = batch_size or self.config.batch_size
 
         val_input_fn, train_input_fn, val_size, val_interval = self.input_pipeline.get_train_input_fns(Xs, Y, batch_size=batch_size)
+        print("Made input fn init")
```
Comment (madisonmay) [inline finetune/base.py]: rogue print

File: finetune/config.py [hunk 7]
```diff
@@ -206,6 +203,7 @@ def get_default_config():
         eval_acc=False,
         save_dtype=None,
         regression_loss="L2",
+        viable_edges=None,
```
Comment (madisonmay) [inline finetune/config.py]: we should talk about what pattern we follow for model specific params, because continuously adding to this config object seems kinda suspect

File: finetune/encoding.py [hunk 8]
```diff
@@ -173,6 +173,8 @@ def _encode(self, texts, labels=None, verbose=True):
         
         for i, text in enumerate(texts):
             if labels is not None:
+                print(labels)
```
Comment (madisonmay) [inline finetune/encoding.py]: rogue prints

File: finetune/input_pipeline.py [hunk 9]
```diff
@@ -217,6 +217,7 @@ def get_train_input_fns(self, Xs, Y=None, batch_size=None, val_size=None):
                     )
         else:
             self.config.dataset_size = len(Xs)
+        print("this shouldnt be a problem")
```
Comment (madisonmay) [inline finetune/input_pipeline.py]: rogue print

File: finetune/input_pipeline.py [hunk 10]
```diff
@@ -241,12 +244,15 @@ def get_train_input_fns(self, Xs, Y=None, batch_size=None, val_size=None):
             Xs_tr, Xs_va, Y_tr, Y_va = train_test_split(Xs, Y, test_size=self.config.val_size, random_state=self.config.seed)
             Xs_tr, Y_tr = self.resampling(Xs_tr, Y_tr)
             self.config.dataset_size = len(Xs_tr)
+            print("before val dataset")
```
Comment (madisonmay) [inline finetune/input_pipeline.py]: you get the idea

File: finetune/network_modules.py [hunk 11]
```diff
@@ -350,3 +351,104 @@ def seq_lab_internal(hidden):
...
+                logits=associations_flat,
+                labels=tf.reshape(targets["associations"], shape=[-1]),
+                weights=tf.reshape(mask, shape=[-1])
+            )
+
+        association_loss = tf.Print(association_loss, [association_loss, -log_likelihood])
+
+        return {
+            'logits': {"sequence": logits, "association": associations},
+            'losses': -log_likelihood + 100 * association_loss,  # TODO: think about weighting.
```
Comment (madisonmay) [inline finetune/network_modules.py]: You already have this in a todo, but yeah definitely worth considering how we intend to make this weighting robust.

File: finetune/network_modules.py [hunk 12]
```diff
@@ -350,3 +351,104 @@ def seq_lab_internal(hidden):
...
+            b = tf.expand_dims(association_head, 2)
+
+            probs = tf.nn.softmax(logits)
+
+            features = tf.concat(
+                [
+                    a - b, a * b,
+                    tf.tile(a, [1, length, 1, 1]),
+                    tf.tile(b, [1, 1, length, 1]),
+                    tf.tile(tf.expand_dims(probs, 1),[1, length, 1, 1]),
```
Comment (madisonmay) [inline finetune/network_modules.py]: Hrmmm ok.... feels weird to directly pass the output of the softmax as features.  I agree we want this information to be accessible by the association head, but think it might warrant some more thought about how to make sure that happens.

File: tests/test_association.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+        data = TreebankNounVP()
+        return (
+            data.get_data(rows),
+            {
+                "possible_associations": ["has_verb"],
+                "viable_edges": {
+                    "noun_phrase": [
+                        ["verb", "has_verb"],
+                    ]
+                    , "verb": [None]
```
[the remark this answers is not in the document]
Comment (benleetownsend) [reply tests/test_association.py]: See the documentation on this in the DocString to the association task, the None has special meaning indicating that a verb is permitted to associate with nothing.
```
"noun_phrase": [["verb", "has_verb"]]
```
says that for every noun_phrase there exists a verb for which the "has_verb" relation holds.

```
"noun_phrase": [["verb", "has_verb"], None]
```
says that for every noun_phrase *may* exist a verb for which the "has_verb" relation holds.

File: finetune/association.py (hunk 1 above)
Comment (benleetownsend) [reply finetune/association.py]: I agree

File: finetune/association.py (hunk 3 above)
Comment (benleetownsend) [reply finetune/association.py]: I think this is something I wanted to discuss, I don't think chunking makes any sense in the case of the association model. So I wanted to remove it entirely, but i'm not sure whether having the restricted maximum separation of associations is something that is okay? So I left it in.

File: finetune/utils.py [hunk 14]
```diff
@@ -333,6 +402,14 @@ def indico_to_finetune_sequence(texts, labels=None, multi_label=True, none_value
                 doc_labels.append([none_value])
             else:
                 doc_labels.append(none_value)
+                doc_association_idx.append(-1)
+                doc_association_type.append(none_value)
+                doc_current_label_idx.append(-2)
```
[the remark this answers is not in the document]
Comment (benleetownsend) [reply finetune/utils.py]: So the doc_current_label_idx is the index of the current label in the original data layout. This final segment by definitition does not have a label and therefore no index. It is given -2 as -1 is already taken as the equivalent placeholder for the doc association_idx. Hopefully this answers your question.

File: finetune/network_modules.py (hunk 11 above)
Comment (madisonmay) [reply finetune/network_modules.py]: Make this configurable

**Comment by madisonmay** [comment]
LGTM, merging
